src/graphs/TopologicalSort.re

```diff
@@ -0,0 +1,75 @@
+type node = {
+    id: string,
+    neighbours: list(string),
+};
+
+type mark =
+  | Temp
+  | Perm;
+
+type directedGraph = list(node);
+
+exception Not_found(string);
+
+exception Graph_not_DAG;
+
+let parseAdjList = adj_list => {
+    let adj_tbl = Hashtbl.create(List.length(adj_list));
+    let insert = ({id, neighbours}) => {
+        Hashtbl.add(adj_tbl, id, neighbours);
+    };
+
+    List.iter(insert, adj_list);
+    let validateNeighbours = node => {
+        List.iter(neighbour => {
+            if (!Hashtbl.mem(adj_tbl, neighbour)) {
+                raise(Not_found(neighbour));
+            }
+        }, node.neighbours);
+    };
+
+    List.iter(validateNeighbours, adj_list);
+    adj_tbl;
+};
+
+let rec visit = (~node_id, ~adj_tbl, ~mark_tbl, ~list) => {
```

Comment:
I'm a bit confused about what happens here. Can you add some comments explaining how `visit` works

Reply:
Sure, I'll add comments